Add QNX support

dchordia opened the pull request:
Add QNX specific changes for libmetal.

Following API additions are made to support libmetal on QNX systems
- allocate and free memory 
- assert statements
- caching support
- condition variables
- physical devices like static shared memory (requires bootup device initialization for usage)
- i/o configurations
- interrupt requests
- logs
- mutex
- dynamic shared memory (under `/dev/shmem`) with precise page-wise controls
- sleep
- time
- system and misc utilities

**NOTE**: Custom QNX RPi4 image is used to perform the tests and report results. The usual QNXE RPi4 image can be used but device.c test maybe expected fail on it, since custom bootup device initialization followed by populating the device with `EXPECTED_ID 0x10` is required.

All tests pass with abovementioned setup
```bash
metal: debug:     registered generic bus
metal: info:      running [atomic]
metal: info:      result [atomic]............................ pass
metal: info:      running [mutex]
metal: info:      result [mutex]............................. pass
metal: info:      running [shmem]
metal: info:      result [shmem]............................. pass
metal: info:      running [condition]
metal: info:      result [condition]......................... pass
metal: info:      running [spinlock]
metal: info:      result [spinlock].......................... pass
metal: info:      running [alloc]
metal: info:      result [alloc]............................. pass
metal: info:      running [irq]
metal: debug:     irq: 0 interrupt associated with irq 1000
metal: debug:     irq: 1 interrupt associated with irq 1001
metal: info:      result [irq]............................... pass
metal: info:      running [device]
metal: info:      result [device]............................ pass
metal: info:      running [version]
metal: info:      result [version]........................... pass
metal: debug:     metal_qnx_irq_shutdown
metal: debug:     unregistered generic bus

```

**Some notes for using this library**
- Multi-page support is not added since current RPi4 has single page at the moment
- Set `static device` virtual address to null for default initialization
- For precise `static device` control, correctly populate `io->physmap` array with page-aligned physical addresses. If not then do not "open" static region greater than current default page size (i.e. 4Kb)
- QNX uses `mem_offset64` calls for getting virtual to physical mapping. `_metal.pagemap_fd` is not required.

dchordia committed:
Add QNX support

y-yukinori commented on lib/system/qnx/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+collect (PROJECT_LIB_SOURCES condition.c)
+collect (PROJECT_LIB_SOURCES device.c)
+collect (PROJECT_LIB_SOURCES init.c)
+collect (PROJECT_LIB_SOURCES irq.c)
```
Thank you for porting libmetal to QNX!
Is io.c also needed here?

dchordia replied:
Hello, yes io.c is missing.
I have made local changes and currently working on unit tests for libmetal, hence the port is not yet complete. Once done it will go through team review on this PR and following which it will be good to use

dchordia committed:
Add tests and library updates

Add tests for the following
- atomic variable
- allocations
- condition variable
- interrupt requests
- mutex
- shared memory
- spinlock
- threads
- shared memory device

y-yukinori commented on lib/system/qnx/alloc.h:
```diff
@@ -0,0 +1,39 @@
+ */
+
+/*
+ * @file	qnx/alloc.c
```
```suggestion
 * @file	qnx/alloc.h
```

y-yukinori commented on lib/system/qnx/assert.h:
```diff
@@ -0,0 +1,27 @@
+#include <assert.h>
+
+/**
+ * @brief Assertion macro for bare-metal applications.
```
```suggestion
 * @brief Assertion macro for QNX-based applications.
```

y-yukinori commented on lib/system/qnx/condition.c:
```diff
@@ -0,0 +1,21 @@
+	int ret = pthread_cond_wait(&cv->cond, m);
+  if (ret == EOK)
+    return 0;
+  return ret;
```
```suggestion
  return -ret;
```

y-yukinori commented on lib/system/qnx/condition.h:
```diff
@@ -0,0 +1,63 @@
+
+static inline void metal_condition_init(struct metal_condition *cv)
+{
+	(cv = cv);
```
```suggestion
	metal_assert(cv);
```

y-yukinori commented on lib/system/qnx/condition.h:
```diff
@@ -0,0 +1,63 @@
+  int ret = pthread_cond_signal(&cv->cond);
+	if (ret == EOK)
+    return 0;
+  return ret;
```
```suggestion
  return -ret;
```

y-yukinori commented on lib/system/qnx/condition.h:
```diff
@@ -0,0 +1,63 @@
+	int ret = pthread_cond_broadcast(&cv->cond);
+  if (ret == EOK)
+    return 0;
+  return ret;
```
```suggestion
  return -ret;
```

y-yukinori commented on lib/system/qnx/device.c:
```diff
@@ -0,0 +1,44 @@
+ */
+
+/*
+ * @file	qnx/sys.h
```
```suggestion
 * @file	qnx/device.c
```

y-yukinori commented on lib/system/qnx/device.c:
```diff
@@ -0,0 +1,44 @@
+
+/*
+ * @file	qnx/sys.h
+ * @brief	QNX system primitives for libmetal.
```
```suggestion
 * @brief	QNX libmetal device operations.
```

y-yukinori commented on lib/system/qnx/init.c:
```diff
@@ -0,0 +1,77 @@
+struct cache_ctrl __qnx_cache_control;
+static int GENERIC_BUS_REGISTER = 0;
+
+void metal_init_page_sizes()
```
```suggestion
static void metal_init_page_sizes(void)
```

y-yukinori commented on lib/system/qnx/init.c:
```diff
@@ -0,0 +1,77 @@
+	ret = metal_bus_register(&metal_generic_bus);
+	if (ret != 0) {
+		metal_log(METAL_LOG_DEBUG, "generic bus init failed - %s\n",
+				strerror(errno));
```
```suggestion
				strerror(-ret));
```

y-yukinori commented on lib/system/qnx/init.c:
```diff
@@ -0,0 +1,77 @@
+	ret = metal_qnx_irq_init();
+	if (ret != 0) {
+		metal_log(METAL_LOG_ERROR, "irq init failed - %s\n",
+				strerror(ret));
```
```suggestion
				strerror(-ret));
```

y-yukinori commented on lib/system/qnx/io.c:
```diff
@@ -0,0 +1,42 @@
+	void *addr;
+  int flags, prot;
+
+  flags = io->mem_flags | MAP_PHYS;
```
```suggestion
  flags = (io->mem_flags & ~PROT_MASK) | MAP_PHYS;
```

y-yukinori commented on lib/system/qnx/io.c:
```diff
@@ -0,0 +1,42 @@
+  int flags, prot;
+
+  flags = io->mem_flags | MAP_PHYS;
+  prot = PROT_READ | PROT_WRITE;
```
```suggestion
  prot = (io->mem_flags & PROT_MASK) | PROT_READ | PROT_WRITE; // PROT_NOCACHE for MMIO allowed. PROT/MAP bits don't overlap.
```

y-yukinori commented on lib/system/qnx/irq.c:
```diff
@@ -0,0 +1,252 @@
+      }
+
+      /* enable receiving of new interrupts */
+      if (InterruptUnmask(0, irq_id) == -1) {
```
```suggestion
      if (InterruptUnmask(irq, irq_id) == -1) {
```

y-yukinori commented on lib/system/qnx/mutex.h:
```diff
@@ -0,0 +1,78 @@
+ */
+
+#ifndef __METAL_MUTEX__H__
+#error "Include metal/mutex.h instead of metal/generic/mutex.h"
```
```suggestion
#error "Include metal/mutex.h instead of metal/qnx/mutex.h"
```

y-yukinori commented on lib/system/qnx/shmem.c:
```diff
@@ -0,0 +1,101 @@
+	ret = mlock(mem, size);
+	if (ret) {
+				metal_log(METAL_LOG_WARNING, "failed to mlock shmem - %s\n",
+			  strerror(-ret));
```
```suggestion
			  strerror(errno));
```

dchordia committed:
update formatting

y-yukinori commented on lib/system/qnx/mutex.h:
```diff
@@ -0,0 +1,78 @@
+
+static inline void __metal_mutex_init(metal_mutex_t *mutex)
+{
+	(mutex = mutex);
```
```suggestion
	metal_assert(mutex);
```

y-yukinori commented on lib/system/qnx/mutex.h:
```diff
@@ -0,0 +1,78 @@
+
+static inline void __metal_mutex_deinit(metal_mutex_t *mutex)
+{
+	(mutex = mutex);
```
```suggestion
	metal_assert(mutex);
```

y-yukinori commented on lib/system/qnx/mutex.h:
```diff
@@ -0,0 +1,78 @@
+
+static inline void __metal_mutex_acquire(metal_mutex_t *mutex)
+{
+	(mutex = mutex);
```
```suggestion
	metal_assert(mutex);
```

y-yukinori commented on lib/system/qnx/mutex.h:
```diff
@@ -0,0 +1,78 @@
+
+static inline void __metal_mutex_release(metal_mutex_t *mutex)
+{
+	(mutex = mutex);
```
```suggestion
	metal_assert(mutex);
```

y-yukinori reviewed (commented):
Thank you for the source code updates!
I’ve been using this libmetal together with OpenAMP for about two months now.
I left some comments on areas that I’ve already fixed locally and a few points that caught my attention.
Sorry for reviewing this PR without being asked.
If you decide that the changes are unnecessary, please feel free to reject them.

dchordia committed:
Add suggested updates

dchordia committed:
remove redunant assign

dchordia commented:
> Thank you for the source code updates! I’ve been using this libmetal together with OpenAMP for about two months now. I left some comments on areas that I’ve already fixed locally and a few points that caught my attention. Sorry for reviewing this PR without being asked. If you decide that the changes are unnecessary, please feel free to reject them.

Thank you for quick and detailed review. Really appreciate it!
I have made the suggested changes except for `(val = val)` and `InterruptUnmask(0, ...)`. Reason for `(val = val)` was libmetal forcing `-Werror -Wall` during compilation. Anyway, I removed them and added allowance for unused parameters in CFLAGS [here](https://github.com/qnx-ports/build-files/pull/218/files#diff-68333f74725a9a31b3077233cc2b88d0997400f44fc4b70b8266190ce49ce63dR56). For 2nd one, can you share reason for passing `irq` to `InterruptUnmask`? Official [documentation](https://www.qnx.com/developers/docs/8.0/com.qnx.doc.neutrino.lib_ref/topic/i/interruptunmask.html?hl=interruptunmask) states it should be set to 0.

Additionally, would you like to raise a PR for QNX port of [open-amp](https://github.com/qnx-ports/open-amp) to add your changes?

y-yukinori commented on lib/system/qnx/io.c:
```diff
@@ -0,0 +1,42 @@
+  int flags, prot;
+
+  flags = (io->mem_flags & ~PROT_MASK) | MAP_PHYS;
+  prot = (io->mem_flags & PROT_MASK) | PROT_READ ;
```
`PROT_WRITE` seems to have been removed.
```suggestion
  prot = (io->mem_flags & PROT_MASK) | PROT_READ | PROT_WRITE;
```

y-yukinori commented:
> Thank you for quick and detailed review. Really appreciate it! I have made the suggested changes except for `(val = val)` and `InterruptUnmask(0, ...)`. Reason for `(val = val)` was libmetal forcing `-Werror -Wall` during compilation. Anyway, I removed them and added allowance for unused parameters in CFLAGS [here](https://github.com/qnx-ports/build-files/pull/218/files#diff-68333f74725a9a31b3077233cc2b88d0997400f44fc4b70b8266190ce49ce63dR56).

Thank you for the quick follow-up and the changes!

> For 2nd one, can you share reason for passing `irq` to `InterruptUnmask`? Official [documentation](https://www.qnx.com/developers/docs/8.0/com.qnx.doc.neutrino.lib_ref/topic/i/interruptunmask.html?hl=interruptunmask) states it should be set to 0.

Sorry, I should have mentioned the environment earlier.
I'm using QNX 7.1, not QNX 8.0.
In this environment, passing `0` as the first argument to `InterruptUnmask` didn’t work, so I had to specify the actual `irq` value instead.
https://www.qnx.com/developers/docs/7.1/#com.qnx.doc.neutrino.lib_ref/topic/i/interruptunmask.html

> Additionally, would you like to raise a PR for QNX port of [open-amp](https://github.com/qnx-ports/open-amp) to add your changes?

Regarding the open-amp PR:
I was able to get it working with just the following two changes, so I don't have a proper PR to contribute.
My local modifications also include some OS-specific code inserted in places where it shouldn't be, so they're not suitable for upstream.
* Avoided a name conflict with QNX's `CACHE_FLUSH` macro in `virtqueue.h`, `remoteproc_virtio.h`, `rpmsg_virtio.h`.
* Added a simple CMake configuration for QNX 7.1.

eleir9268 commented on lib/system/qnx/init.c:
```diff
@@ -0,0 +1,77 @@
+	cache_fini(&__qnx_cache_control);
+	metal_qnx_irq_shutdown();
+	if (GENERIC_BUS_REGISTER)
+		metal_bus_unregister(&metal_generic_bus);
```
A bit nitpicky but `if` without enclosing `{}` can cause issues with multiline macro fns. There might be a MISRA standard for avoiding it.

eleir9268 commented on lib/system/qnx/io.c:
```diff
@@ -0,0 +1,42 @@
+{
+  if (io->virt != NULL)
+    return 0;
+	void *addr;
```
Same as above comment, but there's also an indentation issue.

eleir9268 commented on lib/system/qnx/shmem.c:
```diff
@@ -0,0 +1,101 @@
+		return -ENOMEM;
+	}
+
+	for (virt = mem, page = 0; page < pages; page++) {
```
Again, nitpicky, but `++page` is slightly more efficient than `page++`.

eleir9268 commented on lib/system/qnx/time.c:
```diff
@@ -0,0 +1,29 @@
+
+unsigned long long metal_get_timestamp(void)
+{
+	unsigned long long time = 0;
```
Whitespace issue.

eleir9268 commented on lib/system/qnx/time.c:
```diff
@@ -0,0 +1,29 @@
+  }
+  time = ts.tv_sec * NS_PER_S;
+  time += ts.tv_nsec;
+	return 0;
```
Whitespace issue.

eleir9268 commented on lib/system/qnx/utilities.c:
```diff
@@ -0,0 +1,98 @@
+  flags |= MAP_SHARED;
+
+  if (fd < 0) {
+    fd = -1;
```
Better to use `NOFD`.

eleir9268 commented on lib/system/qnx/utilities.c:
```diff
@@ -0,0 +1,98 @@
+*/
+int metal_unmap(void *mem, size_t size)
+{
+  return munmap(mem, size) < 0 ? -errno : 0;
```
Again, nitpicky, but it's better to use `munmap(mem, size) != 0` so you don't depend on the implementation of the value returned in a fail state.

dchordia committed:
update reviewed changes + whitespace fix across files

dchordia commented:
> Sorry, I should have mentioned the environment earlier. I'm using QNX 7.1, not QNX 8.0. In this environment, passing `0` as the first argument to `InterruptUnmask` didn’t work, so I had to specify the actual `irq` value instead. https://www.qnx.com/developers/docs/7.1/#com.qnx.doc.neutrino.lib_ref/topic/i/interruptunmask.html
> 
Ah interesting, I completely missed this. Should be good now

 > Regarding the open-amp PR:
I was able to get it working with just the following two changes, so I don't have a proper PR to contribute.
>
No worries. Thanks for sharing the updates. I will try those changes

dchordia committed:
more whitespace fixes and return value changes

dchordia commented:
@eleir9268 should licensing have "BlackBerry Limited" or "QNX Software Systems Limited"? I see you have added BlackBerry [here](https://github.com/qnx-ports/check-tools/blob/c5ffa5f4199f10f0d367c5f8633d0c67a6e25a99/check_utils/ctest.py#L2). For my changes I have added QNX one at all places

nnarayanamurthy reviewed (approved):
Hopefully, all the constant values and #define directives are verified, apart from these everything looks fine.

y-yukinori commented on lib/system/qnx/irq.c:
```diff
@@ -0,0 +1,253 @@
+			}
+
+			/* enable receiving of new interrupts */
+      int ret = __QNX__ < 800 ? InterruptUnmask(irq, irq_id) : InterruptUnmask(0, irq_id);
```
`__QNX__` is only used to represent the API version starting from QNX 8.0.
On QNX 7.1, it is evaluated as `1`.

```suggestion
      int ret = __QNX__ >= 800 ? InterruptUnmask(0, irq_id) : InterruptUnmask(irq, irq_id);
```

eleir9268 commented:
> @eleir9268 should licensing have "BlackBerry Limited" or "QNX Software Systems Limited"? I see you have added BlackBerry [here](https://github.com/qnx-ports/check-tools/blob/c5ffa5f4199f10f0d367c5f8633d0c67a6e25a99/check_utils/ctest.py#L2). For my changes I have added QNX one at all places

Good question, I'll ask around for where to find the latest license.

eleir9268 commented:
> > @eleir9268 should licensing have "BlackBerry Limited" or "QNX Software Systems Limited"? I see you have added BlackBerry [here](https://github.com/qnx-ports/check-tools/blob/c5ffa5f4199f10f0d367c5f8633d0c67a6e25a99/check_utils/ctest.py#L2). For my changes I have added QNX one at all places
> 
> Good question, I'll ask around for where to find the latest license.

It looks like 'BlackBerry Limited' is the answer. I'll ping you offline with a documentation link.

dchordia committed:
update licensing

dchordia committed:
fix styling issues

dchordia committed:
code style fixes
